includes/reports/overview/list-table/class-sensei-reports-overview-list-table-courses.php

```diff
@@ -313,4 +319,130 @@ protected function get_additional_filters(): array {
 			'last_activity_date_to'   => $this->get_end_date_and_time(),
 		];
 	}
+
+	/**
+	 * Get total average progress value for courses.
+	 *
+	 * @since  4.5.0
+	 *
+	 * @return float total average progress value for all the courses.
+	 */
+	private function get_total_average_progress(): float {
+		$courses_ids = $this->get_all_courses_ids();
+		if ( empty( $courses_ids ) ) {
+			return false;
+		}
+		$lessons_count_per_courses = $this->get_lessons_in_courses( $courses_ids );
+		$student_count_per_courses = $this->get_students_count_in_courses( $courses_ids );
+		$total_average_progress    = 0;
+
+		foreach ( $courses_ids as $course_id ) {
+			if ( ! isset( $lessons_count_per_courses[ $course_id ] ) || ! isset( $student_count_per_courses[ $course_id ] ) ) {
+				continue;
+			}
+			// Get lessons in the course.
+			$lessons = $lessons_count_per_courses[ $course_id ]->lessons;
+			$lessons = array_map( 'intval', explode( ',', $lessons ) );
+			if ( empty( $lessons ) ) {
+				continue;
+			}
+
+			// Get students count.
+			$students_count = $student_count_per_courses[ $course_id ]->students_count;
+			if ( ! $students_count ) {
+				continue;
+			}
+
+			// Get all completed lessons for all the students.
+			$lesson_args     = array(
+				'post__in' => $lessons,
+				'type'     => 'sensei_lesson_status',
+				'status'   => array( 'graded', 'ungraded', 'passed', 'failed', 'complete' ),
+				'count'    => true,
+			);
+			$completed_count = (int) Sensei_Utils::sensei_check_for_activity( $lesson_args );
```

Comment:
For example, given my database where I have 7-8 courses, where 3 courses have 300 lessons each with 1170 students completing each lesson, the time it is taking to load the page is as follows
![query_with_n+1](https://user-images.githubusercontent.com/6820724/165979296-65de6523-a813-4707-a8f8-70d4d1b89180.png)

As we can see, the time spent in the database is more than 19 seconds. For users with a bigger dataset, it will probably be more, which also will depend on the server config.

For example, instead of fetching in the loop, we can fetch the completion for each lesson from the database in one go, like this for example (need to beautify the query a bit) - 
```
private function get_completions_per_lesson( ): array {

		global $wpdb;
		// phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching -- Safe direct sql.
		return $wpdb->get_results(
			$wpdb->prepare(
				"SELECT wcom.comment_post_id lesson_id, COUNT(*) completion_count
						FROM wp_comments wcom
						WHERE ( wcom.comment_approved = 'graded'
						OR wcom.comment_approved = 'ungraded'
						OR wcom.comment_approved = 'passed'
						OR wcom.comment_approved = 'failed'
						OR wcom.comment_approved = 'complete' )
						AND comment_type IN ('sensei_lesson_status')
						AND wcom.comment_post_ID IN
						(
						select wpm.post_id lesson_id from wp_posts wpc
						join `wp_postmeta` wpm on wpm.meta_value = wpc.id
						where wpm.meta_key = '_lesson_course'
						and wpc.post_status in ('publish','private')
						)
						GROUP BY wcom.comment_post_id"
			),
			'OBJECT_K'
		);
	}
```
This will return a list of lesson_id => completion_count pair. I can see that you are already fetching a list of course_id => lessons pair, so using these two lists, you can easily do a sum and find out the completion for that course.

```
                $lessons_count_per_courses = $this->get_lessons_in_courses( $courses_ids );
		$student_count_per_courses = $this->get_students_count_in_courses( $courses_ids );
		$completions_per_lesson    = $this->get_completions_per_lesson();
		$total_average_progress    = 0;

		foreach ( $courses_ids as $course_id ) {
			...
			...
			...
			//$completed_count = (int) Sensei_Utils::sensei_check_for_activity( $lesson_args );

			$ccompleted_count = array_reduce(
				$lessons,
				function ( $carry, $lesson ) use ( $completions_per_lesson ) {
					$carry += $completions_per_lesson[ $lesson ]->completion_count;
					return $carry;
				},
				0
			);
```
I tried this locally, it was more than 7 seconds faster on average
![query_without_n+1](https://user-images.githubusercontent.com/6820724/165986063-d697f1f6-c3ab-4f5f-8d37-75f767391130.png)